File: Current/Product/Production/Common/Common.Tasks/ApplyFilterTask.vb
Imports NAnt.Core
Imports NAnt.Core.Attributes
Imports System.IO
Imports System



<TaskName("applyfilter")> _
Public Class ApplyFilterTask
    Inherits Task

    Private _Text As TextElement
    Private _Filter As Filters.FilterChain
    Private _OutProperty As String

    <TaskAttribute("propertyname", Required:=True)> _
    Public Property OutProperty() As String
        Get
            Return _OutProperty
        End Get
        Set(ByVal Value As String)
            _OutProperty = Value
        End Set
    End Property


    <BuildElement("text", Required:=True)> _
    Public Property Text() As TextElement
        Get
            Return _Text
        End Get
        Set(ByVal Value As TextElement)
            _Text = Value
        End Set
    End Property

    <BuildElement("filterchain", Required:=True)> _
    Public Property Filter() As Filters.FilterChain
        Get
            Return _Filter
        End Get
        Set(ByVal Value As Filters.FilterChain)
            _Filter = Value
        End Set
    End Property

    Protected Overrides Sub ExecuteTask()
        Me.WriteWithFilters()
    End Sub

    Private Function GetWriter() As TextWriter
        Return New StringWriter()
    End Function

    Private Sub WriteWithFilters()
        Dim Writer As TextWriter
        Dim Reader As StringReader
        Dim ChainedReader As Filters.PhysicalTextReader
        Try
            Writer = Me.GetWriter
            Reader = New StringReader(Me.Text.Value)
            ChainedReader = New Filters.PhysicalTextReader(Reader)
            Dim FilterReader As Filters.Filter = Me.Filter.GetBaseFilter(ChainedReader)
            Do While Not False
                Dim Focus As Integer = FilterReader.Read
                If Not (Focus > -1) Then
                    Exit Do
                End If
                Writer.Write(Microsoft.VisualBasic.ChrW(Focus))
            Loop
            Me.Properties(Me.OutProperty) = Writer.ToString()
        Finally
            If Not Writer Is Nothing Then
                Writer.Close()
            End If
            If Not ChainedReader Is Nothing Then
                ChainedReader.Close()
            End If
            If Not Reader Is Nothing Then
                Reader.Close()
            End If
        End Try
    End Sub

End Class

File: Current/Product/Production/Nant/CIFactory.NAnt.Tasks/Tasks/ApplyFilterTask.cs
using System;
using System.IO;
using NAnt.Core;
using NAnt.Core.Filters;
using NAnt.Core.Attributes;
using CIFactory.NAnt.Types;
using NAnt.Core.Types;

namespace CIFactory.NAnt.Tasks
{
    [TaskName("applyfilter")]
    public class ApplyFilterTask : Task
    {
        #region Fields

        private FilterChain _Filter;

        private string _OutProperty;

        private TextElement _Text;

        #endregion

        #region Properties

        [BuildElement("filterchain", Required = true)]
        public FilterChain Filter
        {
            get { return _Filter; }
            set { _Filter = value; }
        }

        [TaskAttribute("propertyname", Required = true)]
        public string OutProperty
        {
            get { return _OutProperty; }
            set { _OutProperty = value; }
        }

        [BuildElement("text", Required = true)]
        public TextElement Text
        {
            get { return _Text; }
            set { _Text = value; }
        }

        #endregion

        #region Protected Methods

        protected override void ExecuteTask()
        {
            this.WriteWithFilters();
        }

        #endregion

        #region Private Methods

        private TextWriter GetWriter()
        {
            return new StringWriter();
        }

        private void WriteWithFilters()
        {
            TextWriter Writer = null;
            StringReader Reader = null;
            PhysicalTextReader ChainedReader = null;
            try
            {
                Writer = this.GetWriter();
                Reader = new StringReader(this.Text.Value);
                ChainedReader = new PhysicalTextReader(Reader);
                Filter FilterReader = this.Filter.GetBaseFilter(ChainedReader);
                while (!false)
                {
                    int Focus = FilterReader.Read();
                    if (!(Focus > -1))
                    {
                        break; // TODO: might not be correct. Was : Exit Do
                    }
                    Writer.Write((Char)Focus);
                }
                this.Properties[this.OutProperty] = Writer.ToString();
            }
            finally
            {
                if (Writer != null)
                {
                    Writer.Close();
                }
                if (ChainedReader != null)
                {
                    ChainedReader.Close();
                }
                if (Reader != null)
                {
                    Reader.Close();
                }
            }
        }

        #endregion

    }

}